src/openrct2/core/Json.cpp

```diff
@@ -15,53 +15,112 @@
 
 namespace Json
 {
-    json_t* ReadFromFile(const utf8* path, size_t maxSize)
+    json_t ReadFromFile(const utf8* path, size_t maxSize)
     {
-        json_t* json = nullptr;
         auto fs = OpenRCT2::FileStream(path, OpenRCT2::FILE_MODE_OPEN);
 
         size_t fileLength = static_cast<size_t>(fs.GetLength());
         if (fileLength > maxSize)
         {
             throw IOException("Json file too large.");
         }
+        
+        auto fileData = std::string(static_cast<size_t>(fileLength) + 1, '\0');
+        fs.Read(static_cast<void*>(fileData.data()), fileLength);
 
-        utf8* fileData = Memory::Allocate<utf8>(fileLength + 1);
-        fs.Read(fileData, fileLength);
-        fileData[fileLength] = '\0';
+        json_t json;
 
-        json_error_t jsonLoadError;
-        json = json_loads(fileData, 0, &jsonLoadError);
-        Memory::Free(fileData);
-
-        if (json == nullptr)
+        try
+        {
+            json = json_t::parse(fileData, nullptr, true, true);
+        }
+        catch (json_t::exception e)
         {
-            throw JsonException(&jsonLoadError);
+            log_error("Unable to parse JSON file (%s)\n%s", path, e.what());
```

Comment:
This is a bit of a behavior change; . Assuming `JsonExceptions` were uncaught, it would cause termination of the program, whereas yours only logs and moves ahead, supposedly. Not sure if it's in a sane state afterwards.

Reply:
This discussion may be a little out of the scope of this PR, but I've been thinking about the desired behaviour for exceptions since you mentioned it. You're right in that things get a little fuzzy when things don't load (you basically just get a black screen), but do we _always_ want to throw an exception if the JSON couldn't be parsed?

- It makes sense that the core files like rides, etc. would throw an exception as these will never change for 99% of users and anyone changing them would want to fix any problems.
- Do we want exceptions to be thrown for "less important" json files such as the user list or the plugin store?
- Most of the `Object::FromJson` methods for rides rely on having a "properties" key - do we want to throw an exception if this doesn't exist (this would actually make the diffs a bit easier to follow for this PR because I've added indentation for a check on the properties variable).
- We probably _don't_ want exceptions to be thrown by the `Json::FromString`/`Json::FromVector` functions so the game doesn't crash from, e.g. a bad packet.

The From[...] functions will return a `json_t` of type null only if the file couldn't load or the parsing failed, so it would be fairly easy to add an assertion on the returned value.

Reply:
You're right in the sense that we can definitely do better with regards to exception handling, we might not really need to throw every time something fails. That being said, we still want to throw for some stuff and getting the ideal "exception behavior" in this PR might add too much to handle in here.

My suggestions:
- Don't change the exception behavior in here, keep it as is. Let's make this PR about upgrading to modern c++ and doing the changes required for it to work, which will make it easier to review and speed-up the process of merging it.
- You can create an issue with the exact text above and we can all discuss over there on the next steps

Reply:
I've now reverted the above change and a few others so that the exception behaviour should be the same. I've realised that the actual behaviour is not far off from what I've written above.